src/core/sock/sockinfo.h

```diff
@@ -644,4 +649,80 @@ class sockinfo : public socket_fd_api,
     //////////////////////////////////////////////////////////////////
 };
 
+typedef std::unordered_map<flow_tuple, bool> port_no_bind_map_t;
+struct port_map_list {
+    std::list<in_port_t>::iterator it;
+    port_no_bind_map_t map;
+};
+
+class bind_no_port {
+public:
+    /*  Logic of IP_BIND_ADDRESS_NO_PORT
+        In case we need a new port from OS, we call bind with port 0,
+        Otherwise - we call bind with specific port from our DB. */
+    int bind_and_set_port_map(const sock_addr &src, const sock_addr &dst, int fd)
+    {
+        int ret = 0;
+        sock_addr addr(src);
+        socklen_t addr_len = sizeof(addr);
+        std::lock_guard<decltype(m_lock)> lock(m_lock);
```

Comment:
[note] This global lock can become a bottleneck for multithreaded applications in CPS scenario.